k2/csrc/array.h

```diff
@@ -1,45 +1,49 @@
 // k2/csrc/array.h
 
-// Copyright (c)  2020  Xiaomi Corporation (author: Daniel Povey)
+// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey
+//                                                   Haowen Qiu)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
+#include <algorithm>
 #include <functional>
+#include <iterator>
 #include <limits>
 #include <memory>
+#include <type_traits>
+#include <utility>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T, I> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename T, typename I = int32_t>
 struct StridedPtr {
-  T *data;
-  I stride;
-  T &operator[](I i) { return data[i]; }
-  StridedPtr(T *data, I stride) : data(data), stride(stride) {}
-};
-
-/* MIGHT NOT NEED THIS */
-template <typename I, typename Ptr>
-struct Array1 {
-  // Irregular two dimensional array of something, like vector<vector<X> >
-  // where Ptr is, or behaves like, X*.
-  using IndexT = I;
-  using PtrT = Ptr;
-
-  // 'begin' and 'end' are the first and one-past-the-last indexes into `data`
-  // that we are allowed to use.
-  IndexT begin;
-  IndexT end;
-
-  PtrT data;
+  T *data;   // it is NOT owned here
+  I stride;  // in number of elements, NOT number of bytes
+  T &operator[](I i) { return data[i * stride]; }
+  const T &operator[](I i) const { return data[i * stride]; }
+  StridedPtr(T *data = nullptr, I stride = 0)  // NOLINT
+      : data(data), stride(stride) {}
+  StridedPtr(const StridedPtr &other)
```

Comment:
Since `StridedPtr` is trivially constructible and copyable, I think
the default copy constructor and assignment operator provided
by the compiler are good enough and there is no need to implement
our own.

In the constructor `StridedPtr(T *data = nullptr, I stride = 0) `, we should check that

```
DCHECK_GT(stride, 1);
```

Reply:
If we have defined a use-defined constructor, then the compiler will not generate other constructors.  Will add `DCHECK_GT`

Reply:
RE `DCHECK_GT`, as we will declare `PtrT data` without parameter in Array2, or we may write `data = nullptr`, we actually cannot require `stride > 1`.

Reply:
> If we have defined a use-defined constructor, then the compiler will not generate other constructors.

Could you provide some sources for the above comment? I cannot find one. All I've foud is that
> The copy constructor is auto-generated if there is no user-declared move constructor or move assignment operator (because there are no move constructors or move assignment operators in C++03, this simplifies to "always" in C++03) (§12.8/8)

----

I just wrote a test program:

```cpp
#include <iostream>

struct Foo {
  int a;
  int *p;
  Foo(int a = 0, int *p = nullptr) : a(a), p(p) {}
};

int main() {
  int b;
  Foo f(10, &b);
  Foo d;
  d = f;
  std::cout << d.a << "\n";
}
```

The assignment operator of `Foo` is still available even if we have a user defined constructor.

Reply:
> RE DCHECK_GT, as we will declare PtrT data without parameter in Array2, or we may write data = nullptr, we actually cannot require stride > 1.

Instead of writing
```cpp
DCHECK_GT(stride, 1);
```

can we use

```cpp
DCHECK_NE(stride, 1);

```

Reply:
Oh yes, you're right, I just checked https://en.cppreference.com/w/cpp/language/copy_constructor 
https://en.cppreference.com/w/cpp/language/move_constructor
https://en.cppreference.com/w/cpp/language/copy_assignment
Updated, thanks